### PR TITLE
Extract a new Pro_hooks.ml from src/engine/Globals.ml

### PR DESCRIPTION
**Pull request by aryx**
test plan:
make test

### CODE REVIEW

File: src/engine/Pro_hooks.ml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      (** 'pro_hook_ref' is a GADT, this allows us to do an "existential"
+       * quantification over `'a`. That is, you can take any hook with
+       * type '... option ref' and wrap it as a 'pro_hook_ref'. Of course,
+       * when you pattern match on one of this, you cannot make any assumptions
+       * on what that `'a` is, but it is enough to reset the hooks.
+       * So we can enumerate Pro hooks in 'pro_hooks_refs' and then write
+       * 'reset_pro_hooks' and 'save_pro_hooks_and_reset' based on that list.
+       *)
+
+(* TODO: a steps-mode rule to ensure we have all the pro hooks.
```
Comment (aryx) [inline src/engine/Pro_hooks.ml:44]: @emjin steps mode!!! another use case.

Comment (emjin) [reply src/engine/Pro_hooks.ml:44]: Do typed metavariables work for Ocaml? I'm just trying to figure out how to write the part where we detect hooks. (Or we could just assume that they're prefixed with `hook_` lol)

Comment (emjin) [reply src/engine/Pro_hooks.ml:44]: If you write a rule that detects pro hooks I'll turn it multistep for you

**Comment by aryx** [comment]
yes you can assume they start with hook_
any hook_ defined in X.ml, should be also part of this constant in Pro_hooks.ml

**Pull request merged**